File: Baekjoon/Gold4/G4_1647_도시 분할 계획.py
# Union을 사용하여 최소 스패닝 트리를 만든 다음, 스패닝 트리를 구성하는 간선 중 가장 비싼 간선을 뺀다

import sys
import heapq
input = sys.stdin.readline

def find_parents(parents, node):
    while parents[node] != node:
        parents[node] = parents[parents[node]]
        node = parents[node]
    return parents[node]

def Union(parents, node1, node2):
    a = find_parents(parents, node1)
    b = find_parents(parents, node2)
    if a == b:
        return False
    if a < b:
        parents[b] = a
    else:
        parents[a] = b
    return True

def main():
    N, M = map(int, input().split())
    parents = [i for i in range(N+1)]
    edges = []
    for i in range(M):
        A, B, C = map(int, input().split())
        heapq.heappush(edges, (C,A,B))
    weigh = 0
    max_cost = 0
    while edges:
        cost, node1, node2 = heapq.heappop(edges)
        if Union(parents, node1, node2):
            weigh += cost
            max_cost = max(max_cost, cost)
    print(weigh - max_cost)

if __name__ == '__main__':
    main()

File: Baekjoon/Gold4/G4_1647_도시분할계획.java
// 324,940 KB, 1,768 ms
import java.io.*;
import java.util.*;

public class Main {
    static int[] parents;

    static class Node implements Comparable<Node> {
        int start, end, cost;

        public Node(int start, int end, int cost) {
            this.start = start; this.end = end; this.cost = cost;
        }

        @Override
        public int compareTo(Node o) {
            return Integer.compare(this.cost, o.cost);
        }
    }

    static int find_parents(int node){
        while (parents[node] != node){
            parents[node] = parents[parents[node]];
            node = parents[node];
        }
        return parents[node];
    }

    static boolean Union(int node1, int node2){
        int a, b;
        a = find_parents(node1);
        b = find_parents(node2);
        if(a == b) return false;
        if(a < b) parents[b] = a;
        else parents[a] = b;
        return true;
    }

    public static void main(String[] args) throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        parents = new int[N+1];
        for(int i=0; i<N+1; i++) parents[i] = i;
        PriorityQueue<Node> que = new PriorityQueue<>();
        int A, B, C;
        for(int i=0; i<M; i++){
            st = new StringTokenizer(br.readLine());
            A = Integer.parseInt(st.nextToken());
            B = Integer.parseInt(st.nextToken());
            C = Integer.parseInt(st.nextToken());
            que.offer(new Node(A, B, C));
        }

        int weigh = 0;
        int max_cost = 0;
        while(!que.isEmpty()){
            Node node = que.poll();
            if(Union(node.start, node.end)){
                weigh += node.cost;
                max_cost = Math.max(max_cost, node.cost);
            }
        }
        System.out.println(weigh - max_cost);
    }
}